File: brainfuck/bf.pl
use v5.38;
use warnings;
no feature qw(indirect);
use feature qw(signatures);
no warnings qw(experimental);

use Socket;

package Op;

sub new ($class, $op, $val = undef) {
    my $self = {
        op => $op,
        val => $val,
    };
    bless $self, $class;
    return $self;
}


package Tape;

sub new ($class) {
    my $self = {
        tape => [0],
        pos => 0,
    };
    bless $self, $class;
    return $self;
}

sub get ($self) {
    return $self->{tape}[$self->{pos}];
}

sub inc ($self, $x) {
    $self->{tape}[$self->{pos}] += $x;
}

sub move ($self, $x) {
    $self->{pos} += $x;
    my $missing = $self->{pos} - @{$self->{tape}} + 1;
    for (my $i = 0; $i < $missing; $i++) {
        push @{$self->{tape}}, 0;
    }
}


package Printer;

sub new ($class, $quiet) {
    my $self = {
        sum1 => 0,
        sum2 => 0,
        quiet => $quiet
    };
    bless $self, $class;
    return $self;
}

sub print ($self, $n) {
    if ($self->{quiet}) {
        $self->{sum1} = ($self->{sum1} + $n) % 255;
        $self->{sum2} = ($self->{sum2} + $self->{sum1}) % 255;
    } else {
        printf "%c", $n;
    }
}

sub checksum ($self) {
    return ($self->{sum2} << 8) | $self->{sum1};
}


package Main;

my $INC   = 1;
my $MOVE  = 2;
my $PRINT = 3;
my $LOOP  = 4;

sub parse ($source, $i = 0) {
    my $repr = [];
    for (; $i < @{$source}; $i++) {
	my $c = $source->[$i];
        if ($c eq '+') { push @$repr, Op->new($INC,   1); }
        elsif ($c eq '-') { push @$repr, Op->new($INC,  -1); }
        elsif ($c eq '>') { push @$repr, Op->new($MOVE,  1); }
        elsif ($c eq '<') { push @$repr, Op->new($MOVE, -1); }
        elsif ($c eq '.') { push @$repr, Op->new($PRINT); }
        elsif ($c eq '[') {
            my ($parsed_loop, $new_i) = parse($source, $i + 1);
            $i = $new_i;
            push @$repr, Op->new($LOOP, $parsed_loop);
        }
        elsif ($c eq ']') { last; }
    }
    return ($repr, $i);
}

sub run ($parsed, $tape, $p) {
    foreach my $op (@$parsed) {
	my $op_type = $op->{op};
        if ($op_type == $INC)   { $tape->inc($op->{val}); }
        elsif ($op_type == $MOVE)  { $tape->move($op->{val}); }
        elsif ($op_type == $PRINT) { $p->print($tape->get()); }
        elsif ($op_type == $LOOP)  {
            while ($tape->get() > 0) {
                run($op->{val}, $tape, $p);
            }
        }
    }
}

sub notify ($msg) {
    socket(my $socket, Socket::PF_INET, Socket::SOCK_STREAM, (getprotobyname('tcp'))[2]);
    if (connect($socket, Socket::pack_sockaddr_in(9001, Socket::inet_aton('localhost')))) {
        print $socket $msg;
    }
    close($socket);
}

sub verify {
    my $text = "++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>
       ---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.";
    my $p_left = Printer->new(1);
    my @source = split(//, $text);
    my ($parsed, $n) = parse(\@source);
    run($parsed, Tape->new(), $p_left);
    my $left = $p_left->checksum;

    my $p_right = Printer->new(1);
    foreach my $c (split //, "Hello World!\n") {
        $p_right->print(ord($c));
    }
    my $right = $p_right->checksum;

    if ($left != $right) {
        print STDERR "${left} != ${right}\n";
        exit(1);
    }
}

if ($0 eq __FILE__) {
    verify;
    open (my $fh, "<", shift) or die $!;
    undef $/;
    $| = 1;
    read $fh, my $text, -s $fh;
    close($fh);
    my $p = Printer->new($ENV{'QUIET'});

    notify("Perl\t". $$);
    my @source = split(//, $text);
    my ($parsed, $n) = parse(\@source);
    run($parsed, Tape->new(), $p);
    notify("stop");

    if ($p->{quiet}) {
        printf("Output checksum: %d\n", $p->checksum);
    }
}

File: brainfuck/bf.php
<?php

class StringIterator {
    function __construct($str) {
        $this->str = $str;
        $this->current = 0;
        $this->last = strlen($str) - 1;
    }

    function done() {
        return $this->current > $this->last;
    }

    function next() {
        if ($this->current > $this->last)
            throw new Exception("StopIteration");
        else
            return $this->str[$this->current++];
    }
}

class Tape {
    function __construct() {
        $this->pos = 0;
        $this->tape = [0];
    }

    function inc($x) {
        $this->tape[$this->pos] += $x;
    }

    function move($x) {
        $this->pos += $x;
        while ($this->pos >= count($this->tape)) $this->tape []= 0;
    }

    function get() {
        return $this->tape[$this->pos];
    }
}

class Printer {
    function __construct($quiet) {
        $this->sum1 = 0;
        $this->sum2 = 0;
        $this->quiet = $quiet;
    }

    function print($n) {
        if ($this->quiet) {
            $this->sum1 = ($this->sum1 + $n) % 255;
            $this->sum2 = ($this->sum2 + $this->sum1) % 255;
        } else {
            echo chr($n);
        }
    }

    function checksum() {
        return ($this->sum2 << 8) | $this->sum1;
    }
}


class Op {
    function __construct($op, $v) {
        $this->op = $op;
        $this->v = $v;
    }
}

class Brainfuck {
    const INC = 1;
    const MOVE = 2;
    const LOOP = 3;
    const PRINT = 4;

    function __construct($text, $p) {
        $this->ops = $this->parse(new StringIterator($text));
        $this->p = $p;
    }

    function parse($iterator) {
        $res = [];
        while (!$iterator->done()) {
            switch($iterator->next()) {
            case '+': $res []= new Op(self::INC, 1); break;
            case '-': $res []= new Op(self::INC, -1); break;
            case '>': $res []= new Op(self::MOVE, 1); break;
            case '<': $res []= new Op(self::MOVE, -1); break;
            case '.': $res []= new Op(self::PRINT, 0); break;
            case '[': $res []= new Op(self::LOOP, $this->parse($iterator)); break;
            case ']': return $res;
            }
        }
        return $res;
    }

    function _run($ops, $tape) {
        foreach($ops as $op) {
            switch($op->op) {
            case self::INC: $tape->inc($op->v); break;
            case self::MOVE: $tape->move($op->v); break;
            case self::LOOP: while ($tape->get() > 0) $this->_run($op->v, $tape); break;
            case self::PRINT: $this->p->print($tape->get()); break;
            }
        }
    }

    function run() {
        $this->_run($this->ops, new Tape());
    }
}

function notify($msg) {
    $fp = stream_socket_client("tcp://localhost:9001", $errno, $errstr);
    if (!$fp) die("$errstr ($errno)");
    fwrite($fp, $msg);
    fclose($fp);
}

function verify() {
    $text = "++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>
        ---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.";
    $p_left = new Printer(true);
    (new Brainfuck($text, $p_left))->run();
    $left = $p_left->checksum();

    $p_right = new Printer(true);
    foreach(str_split("Hello World!\n") as $c) {
        $p_right->print(ord($c));
    }
    $right = $p_right->checksum();
    assert($left == $right);
}

verify();
$text = file_get_contents($argv[1]);
$p = new Printer(@$_SERVER["QUIET"]);

notify("PHP\t" . getmypid());
(new Brainfuck($text, $p))->run();
notify('stop');

if ($p->quiet) {
    echo "Output checksum: ".$p->checksum()."\n";
}
